### ISSUE TITLE
StaticPathHandler not sending the javascript and css files for multiple viewers

### ISSUE THREAD

**Issue by hash-ir** [google/neuroglancer#338]
Hi,

I have deployed the neuroglancer tornado server behind an Nginx reverse proxy. There is a home page that let us choose the dataset and sends this information back to the tornado server. In tornado, a token handler is created that receives the dataset information, creates a random token and starts a viewer instance with that token. 

The problem is when multiple users simultaneously choose and submit the dataset in the browser, only some viewer instances are successfully launched and others result in a 404 error. For the latter, I see that the viewers are not registered i.e the token and the viewer object is not present in the `Server.viewers` dictionary. This was weird because `register_viewer` is called for each viewer instance. At first, I thought this is due to some blocking effect and therefore, I used an asynchronous decorator together with `ThreadPoolExecutor` in the token handler. But the problem still persisted. 

Recently, I found out that the `StaticPathHandler` is throwing the 404 errors whenever the `viewer_token` is not the same as `Server.token`. I commented the token generation part in the `Server` constructor and just used `if viewer_token not in self.server.viewers:` as a conditional check in `StaticPathHandler`. This led to a confusing behaviour. For the viewer instances that resulted in a 404, the `StaticPathHandler` sends the text/html file alright but not the text/css and application/javascript files. Sometimes the CSS is sent but not the javascript while other times, even the HTML is not sent. And for the viewers that were launched successfully, the handler doesn't throw any error and sends all the web files. 

I am starting the tornado server by launching `start()` in `server.__main__` like this:
```python
if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    add_server_arguments(ap)
    args = ap.parse_args()
    set_server_bind_address(args.host, args.port)

    initialize_server()
    config.dev_environ = args.dev

    start()
    logger.info(f"Neuroglancer server running at {get_server_url()}")
```
With this, I have commented `start()` in the `register_viewer()` and `defer_callback()` methods.

Here is the token handler:
```python
class TokenHandler(BaseRequestHandler): 
    def post(self):
        logger.info("TokenHandler invoked")
        from neuroglancer.config import params
        
        try:
            # get dataset information from the frontend
            params.acquisition = self.get_argument("acq_name")
            params.version = self.get_argument("data_version")
            print(params.acquisition)
            print(params.version)
        
        except tornado.web.MissingArgumentError as e:
            logger.error(e.args[0])
            self.send_error(404)
            return

        token = make_random_token()
        # create a neuroglancer.Viewer() object with the created token
        pf = PropertyFilter(params, ['mi', 'vc', 'sj'], token) 
        logger.info(pf.viewer.get_viewer_url())

        try:
            # redirect to the viewer url
            self.redirect(pf.viewer.get_viewer_url())
        except Exception as e:
            self.send_error(404, message=e.args[0])
            return
```
and the slightly modified static path handler:

```python
class StaticPathHandler(BaseRequestHandler):
    def get(self, viewer_token, path):
        logger.info("StaticPathHandler invoked")
        # if viewer_token != self.server.token and viewer_token not in self.server.viewers:
        if viewer_token not in self.server.viewers:
            logger.error("viewer not registered")
            self.send_error(404)
            return
        try:
            data, content_type = global_static_content_source.get(path)
            logger.warning(f"Data sent by StaticPathHandler {content_type}")
        except ValueError as e:
            self.send_error(404, message=e.args[0])
            return
        self.set_header('Content-type', content_type)
        self.finish(data)
```
Following are the error and output logs for two viewer instances that resulted in a 404:
1. None of the files with any mime-type sent
![image](https://user-images.githubusercontent.com/45996390/131362779-3499971a-9324-43e8-a30f-3a3af248f110.png)

2. HTML, CSS and javascript sent but some other javascript file is missing
![image](https://user-images.githubusercontent.com/45996390/131362749-abb1d5ff-6a0e-4018-a921-7efdfd3aace5.png)

To summarize, I have the following questions in mind:

1. Why the `register_viewer` method is called twice, once in `Viewer` and `_ViewerHelper` constructors?
2. Is a viewer instance coupled with the server in any other way?
3. Why is there a token generation in the `Server` constructor? Consequently, why the conditional check `if viewer_token != self.server.token` is done in `StaticPathHandler` when it is highly likely that the tokens will always be different?

Thanks & regards,
Hashir

**Comment by jbms** [comment]
The issue is that `Server.viewers` is a WeakValueDictionary, which means that as soon as the Viewer object itself is no longer referenced from Python, the viewer will not be registered anymore.

For your purpose, you could solve that by saving a reference to the Viewer objects somewhere else for as long as you want them to remain valid.  If you just add each Viewer object to a global list they will remain valid indefinitely, though eventually the server may run out of memory.  Alternatively you could set up something so that they remain valid for a certain period of time after last use.

In regards to your questions:

> Why the register_viewer method is called twice, once in Viewer and _ViewerHelper constructors?

I think this is just a mistake, but is harmless.

> Is a viewer instance coupled with the server in any other way?

No, but note the weakref issue.

> Why is there a token generation in the Server constructor? Consequently, why the conditional check if viewer_token != self.server.token is done in StaticPathHandler when it is highly likely that the tokens will always be different?

I think this might be a remnant from a previous version before the per-viewer tokens were used, and can probably be removed.

**Comment by hash-ir** [comment]
The `WeakValueDictionary` was indeed creating the issue but I didn't understand why some of the viewer instances were dereferenced as soon as creating them. Anyhow, I am using a normal dict to store the viewers attached to the server and dereferencing them when the connection is closed. 

You mentioned that I could let the viewer instances remain valid for some time after last use. You do not mean after the connection is closed, right? 

And lastly thank you for answering the questions, I have made the changes at my end.

**Comment by jbms** [comment]
With web servers it can be tricky to define exactly what it means for the connection to close.  Normal HTTP requests are stateless, so there is no specific "connection close" event --- hence the reason for an inactivity timeout.  Neuroglancer does maintain a websocket connection to the server, which does have a well-defined close event, but that can sometimes close spuriously and will attempt to reconnect automatically.  I think it would not be desirable to delete the state any time the websocket closes.
